### ISSUE TITLE
sphinx doc search limitations

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc#2967]

**Comment by hjoliver** [comment]
Appears to be a fundamental limitation: https://github.com/sphinx-doc/sphinx/issues/3301

**Comment by kinow** [comment]
I used it only once in Cylc documentation and it worked as I needed to find a specific token. We can switch to Algolia if a more powerful search engine is necessary. Free for non commercial use: https://www.algolia.com/pricing. And used by a bunch of Open Source projects, as well as small experiments (easier to start with Algolia before adopting something like ElasticSearch).

Just in case that's a valid alternative too :+1:

**Comment by hjoliver** [comment]
We could just add a Google "site:" search box for the online version (with some obvious limitations, e.g. broken for a while when new versions go up...)

**Comment by hjoliver** [comment]
Looks like Sphinx itself would need some modifications to allow use with Algolia.

**Comment by hjoliver** [comment]
(This limitation is quite surprising. Sphinx has been around for a while - surely almost anyone would want exact phrase search at the least!)

**Comment by hjoliver** [comment]
@sadielbartholomew - can we still generate a single-page version to allow in-browser Ctrl-F search, as an interim measure?

**Comment by hjoliver** [comment]
Ah yes, I see "make singlehtml" in the Makefile!

**Comment by hjoliver** [comment]
(Motivation for this issue: @jarich reports users at BOM lamenting the loss of the old single-page version, ugly as it was!).

**Comment by sadielbartholomew** [comment]
@hjoliver, yes, that is in theory fairly easy to do with Sphinx, using the [``singlehtml`` builder](https://www.sphinx-doc.org/en/master/usage/builders/#sphinx.builders.singlehtml.SingleFileHTMLBuilder) as you saw referenced. We would need to set that ``builder`` up from within the ``conf.py`` & then the single-page view can be built where appropriate [via ``sphinx-build -b singlehtml``](http://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-b)``. Saying that, the docs for this builder as linked state "obviously this only works with smaller projects", suggesting there may be scaling or structural issues to do this with our significantly-sized docs.

We have, in fact, also had users here at the Met Office express desire for the single page version just for its ease & power of search. In light of this, perhaps we should escalate getting a single page HTML or even PDF version created in the build?

**Comment by sadielbartholomew** [comment]
> (This limitation is quite surprising. Sphinx has been around for a while - surely almost anyone would want exact phrase search at the least!)

Agreed! @oliver-sanders & I were also very puzzled about the poor nature of the search functionality in Sphinx, as noticed during the development of the Rose docs, given the tool is otherwise so powerful. Looking at the GitHub repo, they actually have an [entire label](https://github.com/sphinx-doc/sphinx/labels) dedicated to ``html search``! Hopefully this means the search feature will improve in newer versions.

Other notable issues with Sphinx in-built searching that we observed in the Rose docs, which uses the same Sphinx version as the Cylc docs at least at the moment:

* search results display matched keywords within plain & not formatted text lifted from the RST files, leaving e.g. directive ``.. <name>::`` & title underline e.g. ``==========`` characters in, & also making internal notes e.g. ``TODO`` visible by search:
![rose_search](https://user-images.githubusercontent.com/30274190/53276818-c23bd080-36f8-11e9-91dd-21b180f12d90.png)
* docs version changes during searches are handled poorly (see metomi/rose#2213)

**Comment by sadielbartholomew** [comment]
... and other notable issue with the Cylc docs that we *can't* blame on Sphinx (& instead relate to some issue with my setup from the conversion #2910):

* a consistently useless search result text preview, which is sourcing the same initial text from some template for reasons unknown (as I have not had time to investigate):
![cylc_search](https://user-images.githubusercontent.com/30274190/53276939-55750600-36f9-11e9-9bfe-cfaf4da3f459.png)

**Comment by oliver-sanders** [comment]
I think there is a Sphinx plugin for nicer searching.

**Comment by hjoliver** [comment]
PRs up to generate single-page User Guide, for Ctrl-F browser search, until this issue is resolved.

**Comment by hjoliver** [comment]
@sadielbartholomew 

> a consistently useless search result text preview, ...

This appears to be some weird interaction with GitHub pages, because the search works fine locally for me, with the exact same generated docs.

I'm not sure what the cause is ... the Rose Docs don't have this problem on GitHub Pages (as you noted above).  As an interim measure I've added a warning to the web site documentation page that search result summaries are currently broken.

**Comment by oliver-sanders** [comment]
I think the reason that search is broken on gh-pages is because we are only uploading the `html` directory, search will require the `doctree` directory as well.

This should go away when we copy across the `rose make-docs` infrastructure into Cylc. `rose make-docs` generates the following file structure:

```
doc -> 2019.01.01  # symlink to latest version (named doc for legacy reasons)
index.html         # redirect to doc/html/index.html
2019.01.01/
    index.html     # redirect to html/index.html
    html/
         index.html
         ...
     doctrees/
         index.rst
         ....
     slides/
         index.html
         ...
      pdf/
          cylc-tutorial.pdf
          rose-tutorial.pdf
          rose-documentation.pdf
    
```

So `http://metomi.github.io/rose/` redirects to `http://metomi.github.io/rose/doc/html/index.html`.

**Comment by sadielbartholomew** [comment]
> This should go away when we copy across the rose make-docs infrastructure into Cylc

Just FYI, I'm not sure this is the plan; I think @kinow was planning to get the docs built via the ``setup.py`` as described in https://github.com/cylc/cylc/pull/2910#issuecomment-453884314, though I could be wrong or have missed some discussion on it.

**Comment by sadielbartholomew** [comment]
> This appears to be some weird interaction with GitHub pages, because the search works fine locally for me, with the exact same generated docs.

Ah yes, I have just had a look & the search result text preview is as it should be locally, for me also. The 'Show Source' sidebar function also has always worked locally but not when setup under GitHub pages, so that may be related.

**Comment by oliver-sanders** [comment]
> Just FYI, I'm not sure this is the plan

Copying across the `rose make-docs` functionality and building via `setup.py` aren't mutually exclusive.

> 'Show Source' sidebar function also has always worked locally

Again likely caused by stripping out the doctrees directory

**Comment by oliver-sanders** [comment]
In `bin/cylc-make-docs` we call `sphinx-build` directly:

```bash
sphinx-build -n -b html ./src built-sphinx/
```

Normally you would use the standard Sphinx Makefile which would make the following call

```bash
sphinx-build -b html -d _build/doctrees   . _build/html
```

I might be barking up the wrong tree here but I'm pretty sure that doctrees have something to do with it.

**Comment by hjoliver** [comment]
`sphinx-build --help` says the `-d` option value defaults to `OUTPUTDIR/.doctrees`, which is what we've got.  (And which isn't being stripped out).

**Comment by kinow** [comment]
Very little knowledge of Sphinx, no idea what's the doctrees folder for, but I think it exists in our GitHub pages: https://github.com/cylc/cylc/tree/gh-pages/doc/built-sphinx/.doctrees

**Comment by kinow** [comment]
Searching something like _"batch"_, there are no JS errors on the browser console, but in the network tab, it is possible to see several 404's. Maybe that could be something contributing to the bad results layout?

![image](https://user-images.githubusercontent.com/304786/53596146-fe44c900-3c03-11e9-957b-1a6655e94d00.png)

**Comment by hjoliver** [comment]
Hmm, yes, interesting.  The browser console is impressively useful...

**Comment by kinow** [comment]
This one https://github.com/cylc/cylc.github.io/pull/2 fixed how search results are displayed. But doesn't fix the main issue reported here, which is the limitations for search in Sphinx :+1:
